gptel-context.el

```diff
@@ -163,32 +164,62 @@ context chunk.  This is accessible as, for example:
         (eq buffer-file-coding-system 'no-conversion))
     (file-missing (message "File \"%s\" is not readable." path))))
 
+(defun gptel-context--handle-binary (path)
+  "Add binary file at PATH to context if supported.
+Return PATH if added, nil if ignored."
+  (if-let* (((gptel--model-capable-p 'media))
+            (mime (mailcap-file-name-to-mime-type path))
+            ((gptel--model-mime-capable-p mime)))
+      (prog1 path
+        (cl-pushnew (list path :mime mime)
+                    gptel-context--alist :test #'equal)
+        (message "File \"%s\" added to context." path))
+    (message "Ignoring unsupported binary file \"%s\"." path)
+    nil))
+
+(defun gptel-context--handle-directory (path action)
+  "Process all files in directory at PATH according to ACTION.
+ACTION should be either `add' or `remove'."
+  (cl-block gptel-context--handle-directory
+    (when (eq action 'add)
+      (unless (y-or-n-p (format "Recursively add all files from directory %s? " path))
+        (cl-return-from gptel-context--handle-directory)))
+    (let ((files (directory-files-recursively path "." t)))
+      (mapc (lambda (file)
+              (unless (file-directory-p file)
+                (pcase-exhaustive action
+                  ('add
+                   (if (gptel--file-binary-p file)
+                       (gptel-context--handle-binary file)
+                     (cl-pushnew (list file) gptel-context--alist :test #'equal)
+                     (message "File \"%s\" added to context." file)))
+                  ('remove
+                   (setf (alist-get file gptel-context--alist nil 'remove #'equal) nil)
+                   (message "File \"%s\" removed from context." file)))))
+            files))))
+
 (defun gptel-context-add-file (path)
```

Comment:
We can add an `&optional confirm-dir` argument to this function that the `(interactive ...)` form can set to `t`.

Reply:
Leaving this comment open since I explain why we want to avoid `called-interactively-p` below.